addons/io/org.openhab.io.hueemulation/src/main/java/org/openhab/io/hueemulation/internal/HueEmulationServlet.java

```diff
@@ -0,0 +1,521 @@
+    @Override
+    protected void service(HttpServletRequest req, HttpServletResponse resp) throws ServletException, IOException {
+        String path = req.getRequestURI();
+        logger.trace("{} {}", req.getMethod(), path);
+        setHeaders(resp);
+
+        // UPNP discovery document
+        if (path.equals(PATH + "/discovery.xml")) {
+            apiDiscoveryXML(req, resp);
+            return;
+        }
+
+        // everything is JSON from here
+        resp.setContentType(APPLICATION_JSON);
+
+        // request for API key
+        if (path.equals(PATH)) {
+            if (pairingEnabled) {
+                apiConfig(req, resp);
+            } else {
+                apiServerError(req, resp, HueErrorResponse.UNAUTHORIZED,
+                        "Not Authorized. Pair button must be pressed to add users.");
+            }
+            return;
+        }
+
+        // All other API requests
+        Matcher m = PATH_PATTERN.matcher(path);
+        if (m.matches()) {
+            String userName = m.group(1);
+            boolean lightsReq = m.group(2) != null;
+            String id = m.group(3);
+            boolean stateReq = m.group(4) != null;
+
+            /**
+             * Some devices (Amazon Echo) seem to rely on the bridge to add an unknown user if pairing is on
+             * instead of using the configApi method
+             */
+            if (pairingEnabled) {
+                addUser(userName);
+            } else if (!authorizeUser(userName)) {
+                apiServerError(req, resp, HueErrorResponse.UNAUTHORIZED, "Not Authorized");
+                return;
+            }
+
+            if (stateReq) {
+                /**
+                 * /api/{username}/{lights}/{id}/{state}
+                 */
+                apiState(id, req, resp);
+            } else if (id != null) {
+                /**
+                 * /api/{username}/{lights}/{id}
+                 */
+                apiLight(id, req, resp);
+            } else if (lightsReq) {
+                /**
+                 * /api/{username}/{lights}
+                 */
+                apiLights(req, resp);
+            } else if (userName != null) {
+                /**
+                 * /api/{username}
+                 */
+                apiDataStore(req, resp);
+            } else {
+                apiServerError(req, resp, HueErrorResponse.NOT_AVAILABLE, "Hue Resource Not Availabale");
+            }
+        }
+
+    }
+
+    /**
+     * Hue API call to set the state of a light
+     *
+     * @param id
+     * @param req
+     * @param resp
+     * @throws IOException
+     */
+    private void apiState(String id, HttpServletRequest req, HttpServletResponse resp) throws IOException {
+        if (!req.getMethod().equals(METHOD_PUT)) {
+            apiServerError(req, resp, HueErrorResponse.METHOD_NOT_AVAILABLE, "Only PUT allowed for this resource");
+            return;
+        }
+        try {
+            Gson gson = new Gson();
```

Comment:
`new Gson()` is potentially an expensive operation. You could define `gson` a a field and reuse it throughout your class.